### PR TITLE
Template share selection bug fix

### LINKED ISSUE #7129: bug: Unable to share templates in nations note
### Describe the bug

In the nation's note, there used to be the ability to share templates with other users. That feature is broken,
In the first step of sharing the templates, the desired template would be selected and moved from left to right on the screen.
The screen shows a failure in this step.

### To Reproduce

image

### Expected behavior

The selected list is to move left to right. 

### Client configuration

- Browser: All
- OpenEMR version: 7.0.2
- Operating system:

### PR DESCRIPTION
**Pull request by juggernautsei**
Fixes #7129

#### Short description of what this resolves:
fixes the failed error received when trying to share a template

### CODE REVIEW

**Comment by juggernautsei** [comment]
I have corrected most of the issues with sharing templates. The template is being saved to the new users profile. Now, it needs to be called back when the user needs it. (Note to self)

**Comment by juggernautsei** [comment]
Hi @mdsupport ,
I see you have a note in the custom_template.php file on line 37. What I am failing to see right now is why the system can't retrieve the shared template. 
image

 Even on reloading the entire modal, the frame is empty.
image

The template_context_search.php does not reference the template_users table at all. 

    let allowTemplateWarning = <?php echo $allowTemplateWarning; ?>;
    <?php if (!$isNN) { ?>
        $(function () {
            $('#contextSearch').select2({
                placeholder: <?php echo xlj('Select Template Context'); ?>,
                width: 'resolve',
                theme: 'bootstrap4',
                ajax: {
                    url: top.webroot_url + '/library/ajax/template_context_search.php',
                    data: function (params) {
                        let query = {
                            search: params.term,
                            csrf_token_form: <?php echo js_escape(CsrfUtils::collectCsrfToken()); ?>
                        };
                        return query;
                    },
                    dataType: 'json',
                },
                <?php require($GLOBALS['srcdir'] . '/js/xl/select2.js.php'); ?>
            });

This populates the Context select dropdown.  Template users is not used to select user. If the user is not the one that created the entries. This is where I am stuck. I don't understand the way the template users table is supposed to relate to the

**Comment by juggernautsei** [comment]
When I remove the criteria to fetch the data by user id. The data for the person that the code is shared to appears in the context window.

**Comment by bradymiller** [comment]
hi @juggernautsei , Look like you've sorted out this issue per the comment in your last commit. If so, then code looks good. Just let us know it is testing well for you and then will bring this in.

**Comment by juggernautsei** [comment]
@bradymiller I have tested on three installs including v7.0.1 and it works well on all three.

File: library/ajax/template_context_search.php [hunk 1]
```diff
@@ -37,16 +37,15 @@
...
+    cl2.cl_list_item_long Like ?
 Group By
     cl2.cl_list_item_long
 createQuery;
 
 $search = $_GET['search'];
 $eSearch = "%" . $search . "%";
 $results = [];
-$r = sqlStatementNoLog($cq, array($eSearch, (int)$_SESSION['authUserID']));
+$r = sqlStatementNoLog($cq, array($eSearch));
```
Comment (sjpadgett) [inline library/ajax/template_context_search.php:48]: There's a reason the query includes the creator id, do you know what it is before removing to make this work?
Are you sure that now all user templates aren't exposed?
Does this affect test templates. I doubt it but worth testing to be sure.

I don't want to have to did into this code as last time I worked in it it was working and my clinic uses both text and NN extensively so I don't want to see any more issues I have to get calls on:)
So far they have not had any issues but only had running 702 for 5 business days.

Comment (juggernautsei) [reply library/ajax/template_context_search.php:48]: @sjpadgett I don't have enough testing to answer your questions. I searched the code trying to find why the authUserID was there. I went all the way back to v5. The auth user ID is in v5 but I am too ignorant to see the purpose. In my use case, there is only one person creating the templates. I can tell you that when removed the parameter. The shared template showed to the person that wanted to use it. They don't show up to everyone and only the ones that are selected show to the one that requested the share. I would have to do more digging and trapping to figure out more than that.

**Comment by mdsupport** [comment]
> Hi @mdsupport ,
> I see you have a note in the custom_template.php file on line 37.

The function on line 38 relates purely to emitting html. Intent must have been to make the original code bit modular and extensible.

**Comment by mdsupport** [comment]
> I don't understand the way the template users table is supposed to relate to the

Dropping existing selection filters will fix your use case but also change basic design. Have you tried to look at commit history to find what broke the original functionality?

**Comment by sjpadgett** [comment]
Copy templates aren't being inserted so retrieving by user id won't show is why it looks like it will work.
Also changing the list type may not be appropriate and should be 4....

**Comment by sjpadgett** [comment]
Several of the queries in personalize.php are broken. Select context is one around L-395

**Comment by juggernautsei** [comment]
@mdsupport  "Have you tried to look at commit history to find what broke the original functionality?"
No, there were a lot of changes made from 5 to 6 since 6. It has not had as many changes in the personalize file. 
 
@sjpadgett  "list type may not be appropriate and should be 4"
In our system the list type is not four that is why the fail comes up when trying to select a context to copy. This is where the whole thing started.

It was sure that it would fix my use case but break past. Someone change the list type from 4 to 3 because that is what saves currently.

**Comment by sjpadgett** [comment]
@juggernautsei Are you sure you are using correctly? or am I?
you should be able to select a user and context. move template(4) cat over to right and stick.  The 4 in query is used  to lookup and show the category details in a bottom panel when you select from left panel.
First thing is to fix the query errors. One query is being passed a filter however it is not bound in statement about L-385.

BTW list type is set in stone(hard coded) and 4 is the component details. 3 is category, 2 is context no clue of 1 and 6

Fixing errors is step one

**Comment by sjpadgett** [comment]
@juggernautsei 
I found fix for the original issue. Looks like the list event is also submitting the form so on refresh the list is emptied because not saved.

Add a event.preventDefault() or in tag event in the `function jsub_selected(form, selectFrom, selectedList)` function. Around L-206 of personalize.php.

Remove change to lookup using 3 and completely remove change to my select lookup script i.e put back authUserID.

**Comment by sjpadgett** [comment]
Hi Sherwin, Where are you at with this? I hope you are also testing flow of selecting a context and user when assigning shared templates. There is still a bug with sql queries. The one error from selecting a context is kind of weird and may be a design issue.
Report back if you're seeing error also.

**Comment by sjpadgett** [comment]
`
[04-Jan-2024 12:22:09 America/New_York] SQL Error with statement:query failed: SELECT * FROM customlists WHERE cl_list_type=3 AND cl_deleted=0 AND cl_list_slno NOT IN (SELECT DISTINCT tu_template_id FROM template_users)  AND cl_list_id=? ORDER BY cl_list_id,cl_list_item_long--You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near &#039;? ORDER BY cl_list_id,cl_list_item_long&#039; at line 1==&gt;C:\xampp\htdocs\openemr\library\custom_template\personalize.php at 398:sqlStatement
`

**Comment by sjpadgett** [comment]
@juggernautsei Could I go ahead and take this issue over. There are other bugs that need fixed. Also warnings that I'm sure will turn into errors in PHP 8.4! I'll try to have a PR up today and bring in tomorrow.

**Comment by juggernautsei** [comment]
@juggernautsei not feeling well right now. I will take this up as soon as I am in better condition. Thanks for the insight.

**Comment by sjpadgett** [comment]
No problem for me to go ahead and fix this since I have already fix a bunch of warnings and sql errors. It's not a slight against you but would make my fixes easier to bring in so I can move on to other things.

**Comment by juggernautsei** [comment]
That is a good idea. Carry on.

**Comment by sjpadgett** [comment]
Thanks buddy, hope you feel better. I'll close this and pick up your issue in my PR.

**Pull request closed without merging**